### PR TITLE
Visual Recognition v3 node (first release)

### PR DESCRIPTION
**Pull request by ylecleach**
This is the first release of the new node for Watson Visual Recognition v3 service. It is working on Bluemix and also on Localhost.

### CODE REVIEW

File: services/visual_recognition/v3.js [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+  // the cleanup removes the temp storage, and I am not sure whether 
+  // it should be called here or after alchemy returns and passed
+  // control back to cbdone.
+
+  function performAction(params, feature, cbdone) {
+
+        var visualRecognition = watson.visual_recognition({
+          api_key: apikey,
+          version: 'v3',
+          version_date: '2016-05-19'
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l94)

File: services/visual_recognition/v3.js [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        if ( msg.params != null && msg.params.classifier_ids != null)
+            params['classifier_ids']=msg.params['classifier_ids'];
+        if ( msg.params != null && msg.params.owners != null)
+            params['owners']=msg.params['owners'];
+        if ( msg.params != null && msg.params.threshold != null)
+            params['threshold']=msg.params['threshold'];
+        performAction(params, feature, actionComplete);
+      } else {
+        this.status({fill:'red', shape:'ring', text:'payload is invalid'});          
+        var message ='Payload must be either an image buffer or a string representing a url';
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Variable 'message' is already declared in line 148.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l328)

File: services/visual_recognition/v3.js [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+// Watson Visual Recognition functions supported by this node
+var FEATURE_RESPONSES = {
+  classifyImage: 'classifyImage',
+  detectFaces: 'detectFaces',
+  recognizeText: 'recognizeText',
+  createClassifier: 'createClassifier',
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l22)

File: services/visual_recognition/v3.js [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+      visualRecognition.detectFaces(params, cbdone);
+    } else if (feature === 'recognizeText') {
+      visualRecognition.recognizeText(params, cbdone);
+    } else if (feature === 'createClassifier') {
+      visualRecognition.createClassifier(params, cbdone);
+    } else if (feature === 'retrieveClassifiersList') {
+      visualRecognition.listClassifiers(params, cbdone);
+    } else if (feature === 'retrieveClassifierDetails') {
+      visualRecognition.getClassifier(params, cbdone);
+    } else if (feature === 'deleteClassifier') {
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l110)

File: services/visual_recognition/v3.js [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+  recognizeText: 'recognizeText',
+  createClassifier: 'createClassifier',
+  retrieveClassifiersList: 'retrieveClassifiersList',
+  retrieveClassifierDetails: 'retrieveClassifierDetails',
+  deleteClassifier: 'deleteClassifier'
+};
+
+module.exports = function (RED) {
+  var cfenv = require('cfenv');
+  var watson = require('watson-developer-cloud');
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l30)

File: services/visual_recognition/v3.js [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+  var imageType = require('image-type');
+  var url = require('url');
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l33)

File: services/visual_recognition/v3.js (hunk 1 above)
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'version_date' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l94)

File: services/visual_recognition/v3.js [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+  var temp = require('temp');
+  var fileType = require('file-type');
+  var fs = require('fs');
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l36)

File: services/visual_recognition/v3.js [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+  };
+
+  // Utility function that performs the Watson Visual Recognition call. 
+  // the cleanup removes the temp storage, and I am not sure whether 
+  // it should be called here or after alchemy returns and passed
+  // control back to cbdone.
+
+  function performAction(params, feature, cbdone) {
+
+        var visualRecognition = watson.visual_recognition({
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 8.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l91)

File: services/visual_recognition/v3.js [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+// Watson Visual Recognition functions supported by this node
+var FEATURE_RESPONSES = {
+  classifyImage: 'classifyImage',
+  detectFaces: 'detectFaces',
+  recognizeText: 'recognizeText',
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686218611/issues/source?bid=3330694&fileBranchId=3336646#l21)

File: services/visual_recognition/v3.js [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+  function imageCheck(data) {
+    return data instanceof Buffer && imageType(data) !== null;
+  };
+
+  function urlCheck(str) {
+    var parsed = url.parse(str)
+    return (!!parsed.hostname && !!parsed.protocol && str.indexOf(' ') < 0);
+  };
+
+  function stream_buffer(file, contents, cb) {
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'stream_buffer' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l77)

File: services/visual_recognition/v3.js (hunk 2 above)
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Variable 'message' is already declared in line 148.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l328)

File: services/visual_recognition/v3.js [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+      var params = {};
+
+      // This is the callback after the call to the watson service.    
+      // Set up as a var within this scope, so it has access to node, msg etc. 
+      // in preparation for the Watson service action       
+      var actionComplete = function(err, body, other) {
+
+        if (err != null && body==null)
+        {
+          node.status({fill:'red', shape:'ring', text:'call to watson visual recognition v3 service failed'});
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l162)

File: services/visual_recognition/v3.js [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+            if (prop.indexOf('_examples')>=0)
+            {
+              // before pushing the function into the task array wrap the push in an IIFE function, passing in the 'prop' parameter
+              (function(prop, list_params, msg) {
+
+               asyncTasks.push(function (callback) {
+                  var buffer = msg.params[prop];
+                  temp.open({suffix: '.' + fileType(buffer).ext}, function (err, info) {
+                    if (err) {
+                      this.status({fill:'red', shape:'ring', text:'unable to open image stream'});
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l273)

File: services/visual_recognition/v3.js [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+            {
+              console.log("Parallel ended with error " + error);
+              return;
+            }
+            for (p in list_params)
+              params[p]=list_params[p];
+            performAction(params, feature, actionComplete2);
+          });
+      }
+        else if (feature=='retrieveClassifiersList') {
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l306)

File: services/visual_recognition/v3.js [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+        node.error(message, msg);
+        return;        
+      }
+
+    });
+  }
+
+  RED.nodes.registerType('visual-recognition-v3', WatsonVisualRecognitionV3Node, {
+    credentials: {
+      apikey: {type:"password"}
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l338)

File: services/visual_recognition/v3.js [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+               asyncTasks.push(function (callback) {
+                  var buffer = msg.params[prop];
+                  temp.open({suffix: '.' + fileType(buffer).ext}, function (err, info) {
+                    if (err) {
+                      this.status({fill:'red', shape:'ring', text:'unable to open image stream'});          
+                      var message ='Node has been unable to open the image stream'; 
+                      node.error(message, msg);
+                      callback('open error on '+prop);
+                    }  
+                    stream_buffer(info.path, msg.params[prop], function () {
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'stream_buffer' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l278)

File: services/visual_recognition/v3.js [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+
+          async.parallel(asyncTasks, function(error, results){
+            // when all temp local copies are ready, copy of all parameters and request to watson api
+            console.log(error, results);
+            if (error)
+            {
+              console.log("Parallel ended with error " + error);
+              return;
+            }
+            for (p in list_params)
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'list_params' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l301)

File: services/visual_recognition/v3.js [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+      }
+      if (feature != 'retrieveClassifiersList' && feature != 'retrieveClassifierDetails' && feature != 'deleteClassifier')
+      if (typeof msg.payload == "boolean" || typeof msg.payload == "number") {
+          this.status({fill:'red', shape:'ring', text:'bad format payload'}); 
+          message = 'Bad format : msg.payload must be a URL string or a Node.js Buffer';
+          node.error(message, msg);
+          return;
+        }
+
+      // If it is present the newly provided user entered key takes precedence over the existing one.
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l142)

File: services/visual_recognition/v3.js [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+          return;
+        }
+        else {
+          msg.result = {};
+          msg.result['all'] = body;
+          msg.payload='see msg.result'; // to remove any Buffer that could remains
+          node.send(msg); 
+          node.status({});
+        }
+      } // actionComplete2
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2686279757/issues/source?bid=3330694&fileBranchId=3336646#l212)

File: services/visual_recognition/v3.js (hunk 2 above)
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Variable 'message' is already declared in line 148.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l328)

File: services/visual_recognition/v3.js [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+            return;        
+          }  
+          stream_buffer(info.path, msg.payload, function () {
+          params['images_file'] = fs.createReadStream(info.path);
+          if ( msg.params != null && msg.params.classifier_ids != null)
+            params['classifier_ids']=msg.params['classifier_ids'];
+          if ( msg.params != null && msg.params.owners != null)
+            params['owners']=msg.params['owners'];
+          if ( msg.params != null && msg.params.threshold != null)
+            params['threshold']=msg.params['threshold'];
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:265]: Codacy Issue found: [Expected '{' and instead saw 'params'.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l253)

File: services/visual_recognition/v3.js [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+            }
+            for (p in list_params)
+              params[p]=list_params[p];
+            performAction(params, feature, actionComplete2);
+          });
+      }
+        else if (feature=='retrieveClassifiersList') { 
+          performAction(params, feature, actionComplete2);
+      }
+        else if (feature =='retrieveClassifierDetails') {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l309)

File: services/visual_recognition/v3.js [hunk 21]
```diff
@@ -0,0 +1,348 @@
...
+        else {
+          msg.result = {};
+          msg.result['all'] = body;
+          msg.payload='see msg.result'; // to remove any Buffer that could remains
+          node.send(msg); 
+          node.status({});
+        }
+      } // actionComplete2
+
+      var actionCompleteDeleteClassifier = function(err, body, other) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:226]: Codacy Issue found: ['other' is defined but never used.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l214)

File: services/visual_recognition/v3.js [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+          });
+      }
+        else if (feature=='retrieveClassifiersList') { 
+          performAction(params, feature, actionComplete2);
+      }
+        else if (feature =='retrieveClassifierDetails') {
+          params['classifier_id']=msg.params['classifier_id'];
+          performAction(params, feature, actionComplete2);
+      }
+        else if (feature=='deleteClassifier') {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:327]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l313)

File: services/visual_recognition/v3.js [hunk 23]
```diff
@@ -0,0 +1,348 @@
...
+      temp.cleanup(); // so there is at most 1 temp file at a time (did not found a better solution...)
+
+      if (!msg.payload) {
+        this.status({fill:'red', shape:'ring', text:'missing payload'}); 
+        message = 'Missing property: msg.payload';
+        node.error(message, msg);
+        return;
+      }
+      if (feature != 'retrieveClassifiersList' && feature != 'retrieveClassifierDetails' && feature != 'deleteClassifier')
+      if (typeof msg.payload == "boolean" || typeof msg.payload == "number") {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l135)

File: services/visual_recognition/v3.js (hunk 12 above)
Comment (yrezgui) [inline services/visual_recognition/v3.js:286]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l273)

File: services/visual_recognition/v3.js [hunk 24]
```diff
@@ -0,0 +1,348 @@
...
+
+  function performAction(params, feature, cbdone) {
+
+        var visualRecognition = watson.visual_recognition({
+          api_key: apikey,
+          version: 'v3',
+          version_date: '2016-05-19'
+        });
+
+    if (feature == 'classifyImage')
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l97)

File: services/visual_recognition/v3.js [hunk 25]
```diff
@@ -0,0 +1,348 @@
...
+  };
+
+  function urlCheck(str) {
+    var parsed = url.parse(str)
+    return (!!parsed.hostname && !!parsed.protocol && str.indexOf(' ') < 0);
+  };
+
+  function stream_buffer(file, contents, cb) {
+    fs.writeFile(file, contents, function (err) {
+      if (err) throw err;
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:79]: Codacy Issue found: [Expected '{' and instead saw 'throw'.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l79)

File: services/visual_recognition/v3.js [hunk 26]
```diff
@@ -0,0 +1,348 @@
...
+
+  // This is the Watson Visual Recognition V3 Node
+  function WatsonVisualRecognitionV3Node (config) {
+    RED.nodes.createNode(this, config);
+    var node = this;
+    var message;
+    this.on('input', function (msg) 
+    {
+      // Check which single feature has been requested.
+      var feature = config["image-feature"];
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2686950655/issues/source?bid=3225115&fileBranchId=3338101#l124)

File: services/retrieve_and_rank/v1.js [hunk 27]
```diff
@@ -51,20 +56,20 @@ module.exports = function (RED) {
         var params = {
           training_data: msg.payload
         };
-      
+
         if (config.rankername) {
-          params.training_metadata = "{\"name\": \""+config.rankername+"\"}";
+          params.training_metadata = '{\"name\": \"'+config.rankername+'\"}';
```
Comment (chughts) [inline services/retrieve_and_rank/v1.js:61]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l61)

File: services/retrieve_and_rank/v1.js [hunk 28]
```diff
@@ -51,20 +56,20 @@ module.exports = function (RED) {
...
         }
-        node.status({fill:"blue",shape:"ring",text:"Uploading training data"});
+        node.status({fill:'blue',shape:'ring',text:'Uploading training data'});
         retrieve_and_rank.createRanker(params, function(err, res) {
             handleWatsonCallback(null,node,msg,err,res,function() {
-              node.status({fill:"blue",shape:"ring",text:"Training data uploaded. Ranker is training"});
+              node.status({fill:'blue',shape:'ring',text:'Training data uploaded. Ranker is training'});
```
Comment (chughts) [inline services/retrieve_and_rank/v1.js:66]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l66)

File: services/retrieve_and_rank/v1.js (hunk 27 above)
Comment (chughts) [inline services/retrieve_and_rank/v1.js:61]: Codacy Issue found: [Identifier 'training_metadata' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l61)

File: services/retrieve_and_rank/v1.js [hunk 29]
```diff
@@ -51,20 +56,20 @@ module.exports = function (RED) {
...
           training_data: msg.payload
         };
-      
+
         if (config.rankername) {
-          params.training_metadata = "{\"name\": \""+config.rankername+"\"}";
+          params.training_metadata = '{\"name\": \"'+config.rankername+'\"}';
         }
-        node.status({fill:"blue",shape:"ring",text:"Uploading training data"});
+        node.status({fill:'blue',shape:'ring',text:'Uploading training data'});
```
Comment (chughts) [inline services/retrieve_and_rank/v1.js:63]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l63)

File: services/retrieve_and_rank/v1.js [hunk 30]
```diff
@@ -542,7 +549,7 @@ module.exports = function (RED) {
 
               default:
                 node.status({});
-                var message = "Ranker training status: "+res.status+", description: "+res.status_description;
+                var message = 'Ranker training status: '+res.status+', description: '+res.status_description;
```
Comment (chughts) [inline services/retrieve_and_rank/v1.js:552]: Codacy Issue found: [Variable 'message' is already declared in line 534.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l552)

File: services/retrieve_and_rank/v1.js [hunk 31]
```diff
@@ -578,7 +585,7 @@ module.exports = function (RED) {
 
                 default:
                   node.status({});
-                  var message = "Cluster status: "+res.status+", description: "+res.status_description;
+                  var message = 'Cluster status: '+res.status+', description: '+res.status_description;
```
Comment (chughts) [inline services/retrieve_and_rank/v1.js:588]: Codacy Issue found: [Variable 'message' is already declared in line 569.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l588)

File: services/retrieve_and_rank/v1.js [hunk 32]
```diff
@@ -292,7 +301,7 @@ module.exports = function (RED) {
               //Note. temporary workaround until bug is fixed in Node SDK
               //Stream zip file from watson api to temp directory,
               //read from temp directory and pass on in msg.payload as buffer
-              var url = "https://gateway.watsonplatform.net/retrieve-and-rank/api/v1/solr_clusters/"+params.cluster_id+"/config/"+params.config_name;
+              var url = 'https://gateway.watsonplatform.net/retrieve-and-rank/api/v1/solr_clusters/'+params.cluster_id+'/config/'+params.config_name;
```
Comment (chughts) [inline services/retrieve_and_rank/v1.js:304]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l304)

File: services/retrieve_and_rank/v1.js (hunk 27 above)
Comment (chughts) [inline services/retrieve_and_rank/v1.js:61]: Codacy Issue found: [Identifier 'training_metadata' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l61)

File: services/retrieve_and_rank/v1.js (hunk 28 above)
Comment (chughts) [inline services/retrieve_and_rank/v1.js:66]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l66)

File: services/retrieve_and_rank/v1.js (hunk 32 above)
Comment (chughts) [inline services/retrieve_and_rank/v1.js:304]: Codacy Issue found: [Identifier 'cluster_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2690663401/issues/source?bid=3330694&fileBranchId=3336646#l304)

**Comment by ylecleach** [comment]
I have fixed all codacy issues. waiting analysis to be completed.

**Pull request closed without merging**